docs/adr/0027-add-objects-api-endpoint.md

```diff
@@ -0,0 +1,108 @@
+---
+# Add an objects API endpoint
+
+## Context and Problem Statement
+
+There is a requirement to identify all Flows that reference (via a Flow Segment) a particular media object.
+This for example supports a business or legal requirement to delete a media object from the store, which requires knowing which Flows reference it.
+The media object can be deleted by deleting all (Flow Segment) references to it.
+
+The proposal is to add a `/objects?object_id={object-id}` endpoint that provides a list of Flows that reference (via Flow Segments) the media object.
+
+## Considered Options
+
+* Option 1: Use the existing endpoints to discover which Flows reference a media object
+* Option 2: Add a `/objects?object_id={object-id}` endpoint that lists all Flows that reference a media object
+* Option 3a: Extend 2 to optionally signal which Flow first referenced the media object in the TAMS instance
+* Option 3b: Restrict 3a to require signalling which Flow referenced the media object first
+* Option 3c: Allow a user to set `first_referenced_by_flow`
+* Option 4: Add an endpoint to list media objects for a Flow
+
+## Decision Outcome
+
+Chosen option: Option 3a, because it is a more efficient way to discover which Flows reference a media object and knowing the first reference may help trace the origin of the media object.
+Is there a need to always have the information available and therefore make the property required, i.e. choose option 3b?
+
+### Implementation
+
+The specification changes have been implemented in PR [#111](https://github.com/bbc/tams/pull/111).
+
+## Pros and Cons of the Options
+
+### Option 1: Use the existing endpoints to discover which Flows reference a media object
+
+The Flows (and Flow Segments) referencing a media object can be found by using a GET of `/flows/{flow-id}/segments?object_id={object-id}` for every Flow.
+
+* Good, because there are no changes to the API
+* Bad, because it requires a client to loop through all the Flows to GET the Flow Segments that reference the media object
+* Bad, because it is likely more efficient for TAMS to be asked to run a single query across all Flows then to run multiple queries for each Flow in turn
+
+### Option 2: Add a `/objects?object_id={object-id}` endpoint that lists all Flows that reference a media object
+
+This `/objects?object_id={object-id}` endpoint provides JSON that lists of all Flows referencing a media object.
+The `referenced_by_flows` property contains the list of Flow IDs.
+
+The Flow Segments that reference the media object can all be found using a GET on `/flows/{flow-id}/segments?object_id={object-id}` for every Flow in the listing rather than every Flow.
+The Flow Segments that reference the media object can all be deleted using a DELETE on `/flows/{flow-id}/segments?object_id={object-id}` for every Flow in the listing.
+
+The `object_id` is as a query parameter rather than a URL path component because the media object ID structure is not restricted.
```

Comment:
Should we consider restricting the `object_id` structure so this could be a path parameter? It would be a breaking change. And it would probably have other implications. But it would allow this part of the API to be much more consistent.

Reply:
The main issue are the slashes as they are expected to also be used to form hierarchies in object storage. Adding them as path components with the slashes would be messy. I guess we could restrict the `object_id` to a UUID string as well for example, but as you say, there may be other implications.

Reply:
Worth still adding it as an option so we've recorded why we rejected it?

Reply:
I've decided to make it a path component - see https://github.com/bbc/tams/pull/111#issuecomment-2704295689